src/indieweb_utils.py

```diff
@@ -42,21 +51,16 @@ def canonicalize_url(url, domain, full_url=None, protocol="https"):
     """
 
     if url.startswith("http://") or url.startswith("https://"):
-        domain = url.split("/")[2]
-
-        # remove port from domain
-
-        domain = domain.split(":")[0]
-        protocol = url.split("/")[0]
-
-        return protocol + "//" + domain + "/" + "/".join(url.split("/")[3:])
-
-    if ":" in domain:
-        text_before_port = domain.split(":")[0]
+        domain = url_parse.urlsplit(url).netloc
+        url_protocol = url_parse.urlsplit(url).scheme
+        path = url_parse.urlsplit(url).path
 
-        text_after_port = domain.split(":")[1].split("/")[0]
+        if url.startswith("http://"):
+            domain = domain.replace(":80", "")
+        elif url.startswith("https://"):
+            domain = domain.replace(":443", "")
 
-        domain = text_before_port + "/" + text_after_port
+        return url_protocol + "://" + domain + path
```

Comment:
Using an f-string might make this easier to read.

```suggestion
        return f"{url_protocol}://{domain}{path}"
```